### PR TITLE
swarm, p2p/protocols: Stream accounting

### PR DESCRIPTION
**Pull request by holisticode**
Introduces pricing for the stream protocol.

This PR finalizes and closes https://github.com/ethersphere/go-ethereum/issues/611.
It also closes https://github.com/ethersphere/go-ethereum/pull/1026 where the basic discussions about the code took place.

PR https://github.com/ethereum/go-ethereum/pull/17951 introduced the lowest level of accounting (protocol-agnostic) to the `p2p/protocols` layer and the necessary interfaces.

PR https://github.com/ethereum/go-ethereum/pull/18050 added implementations for the `swarm` package of the accounting interfaces. Still, nothing would actually be accounted.

This PR finally concludes phase 1 of swap accounting by introducing a price for message exchanges requiring accounting in the `stream` protocol, which is how swarm coordinates retrieval requests and chunk delivery.

The prices in this PR are purely **arbitrary** (1 `wei`) and don't represent any cryptoeconomic metric nor have been backed by any experimental networks nor trials. As Swarm becomes mature, the prices will have to be adjusted and fine-tuned according to usage scenarios and incentives design.

### CODE REVIEW

File: p2p/protocols/accounting.go [hunk 1]
```diff
@@ -114,18 +116,18 @@ func NewAccounting(balance Balance, po Prices) *Accounting {
 //at the passed interval writes the metrics to a LevelDB
```
Comment (justelad) [inline p2p/protocols/accounting.go]: I believe that `godoc` convention is to have `// ...` (//<space><funcName>...)

File: swarm/network/stream/stream.go [hunk 2]
```diff
@@ -235,10 +235,14 @@ func NewRegistry(localID enode.ID, delivery *Delivery, syncChunkStore storage.Sy
 	return streamer
 }
 
+//This is an accounted protocol, therefore we need to provide a pricing Hook to the spec
```
Comment (justelad) [inline swarm/network/stream/stream.go]: comments should be //<space>...

File: p2p/protocols/accounting.go (hunk 1 above)
Comment (holisticode) [reply p2p/protocols/accounting.go]: It doesn't look like it's explicit, but they themselves always DO use a space....

I don't really like it but if that's the convention I will apply from now on

File: swarm/network/stream/stream.go [hunk 3]
```diff
@@ -110,7 +110,7 @@ func NewRegistry(localID enode.ID, delivery *Delivery, syncChunkStore storage.Sy
 	if options.SyncUpdateDelay <= 0 {
 		options.SyncUpdateDelay = 15 * time.Second
 	}
-	//check if retriaval has been disabled
+	// check if retriaval has been disabled
```
Comment (justelad) [inline swarm/network/stream/stream.go]: typo (also i think the comment is not necessary, the code explains itself)

File: p2p/protocols/accounting.go [hunk 4]
```diff
@@ -108,87 +110,87 @@ func NewAccounting(balance Balance, po Prices) *Accounting {
...
+	// do the accounting
 	err := ah.Add(costToLocalNode, peer)
-	//record metrics: just increase counters for user-facing metrics
+	// record metrics: just increase counters for user-facing metrics
 	ah.doMetrics(costToLocalNode, size, err)
 	return err
 }
 
-//Implement Hook.Receive
+// Implement Hook.Receive
```
Comment (justelad) [inline p2p/protocols/accounting.go]: remove the comments that denote that it implements an interface method, comment them on the struct instead saying the `Accounting implements the Hook interface`. the same goes for the `Send` method.

File: p2p/protocols/accounting.go (hunk 1 above)
Comment (justelad) [reply p2p/protocols/accounting.go]: i believe this is the convention across the codebase too

**Comment by zelig** [comment]
@holisticode exactly

**Comment by holisticode** [comment]
@zelig updated with a unit test

File: swarm/network/stream/streamer_test.go [hunk 5]
```diff
@@ -921,3 +921,31 @@ func TestMaxPeerServersWithoutUnsubscribe(t *testing.T) {
 		}
 	}
 }
+
+func TestRetrievalIsBilled(t *testing.T) {
```
Comment (zelig) [inline swarm/network/stream/streamer_test.go]: remove?

File: swarm/network/stream/streamer_test.go [hunk 6]
```diff
@@ -921,3 +921,31 @@ func TestMaxPeerServersWithoutUnsubscribe(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if r.prices == nil {
+		t.Fatal("No prices implementation available for the stream protocol")
+	}
+
+	price := r.prices.Price(&ChunkDeliveryMsgRetrieval{})
+	if price == nil || price.Value == 0 {
```
Comment (zelig) [inline swarm/network/stream/streamer_test.go]: why not check the concrete price set?

File: swarm/network/stream/streamer_test.go [hunk 7]
```diff
@@ -939,13 +938,17 @@ func TestHasPriceImplementation(t *testing.T) {
...
 	}
 
-	price := r.prices.Price(&ChunkDeliveryMsgRetrieval{})
-	if price == nil || price.Value == 0 {
+	pricesInstance, ok := r.prices.(*StreamerPrices)
+	if !ok {
+		t.Fatal("`Registry` does not have the expected Prices instance")
+	}
+	price := pricesInstance.Price(&ChunkDeliveryMsgRetrieval{})
+	if price == nil || price.Value == 0 && price.Value != pricesInstance.getChunkDeliveryMsgRetrievalPrice() {
```
Comment (zelig) [inline swarm/network/stream/streamer_test.go]: this is not right :)

**Comment by hungbui20384** [comment]
Ok

**Pull request merged**